api/lightning_utils.py

```diff
@@ -0,0 +1,189 @@
+import requests, json
+import models as m
+import time
+from extensions import db
+from flask import current_app as app
+
+auth_header = ''
+lndhub_url = ''
+
+class LightningInvoiceUtil:
+    def __init__(self):
+        self.lndhub_url = app.config['LNDHUB_URL']
+        app.logger.info(f"init - LNDHUB_URL={self.lndhub_url}")
+        self.auth_header = self.get_login_token()
+
+    def get_login_token(self):
+        app.logger.info(f"get_login_token - Trying login to LndHub API ({self.lndhub_url})...")
+
+        payload = {
+            'login': app.config['LNDHUB_USER'],
+            'password': app.config['LNDHUB_PASSWORD']
+        }
+
+        r = requests.post(
+            self.lndhub_url + '/auth',
+            data=payload
+        )
+
+        json_res = r.json()
+        app.logger.debug(f"get_login_token - json_res = ({json_res})...")
+
+        if 'error' in json_res:
+            app.logger.error(f"get_login_token - {json_res['message']}")
+            time.sleep(60)
+            return False
+
+        access_token = json_res['access_token']
+        app.logger.info(f"get_login_token - Access token = {access_token}")
+
+        headers = {"Authorization": "Bearer " + access_token}
+
+        return headers
+
+    def create_invoice(self, order_id, sats):
+        payload = {
+            'amount':sats,
+            'description':'Payment for Order #' + str(order_id)
+        }
+        app.logger.debug(f"create_invoice - Creating invoice for order: {payload}")
+
+        response_invoice = requests.post(
+            self.lndhub_url + '/v2/invoices',
+            headers = self.auth_header,
+            json=payload
+        )
+
+        if (response_invoice.status_code == 200):
+            app.logger.info(f"create_invoice - The request to create a new LN invoice was a success!")
+
+            json_response_invoice = response_invoice.json()
+            app.logger.debug(f"create_invoice - The request to create a new LN invoice was a success!")
+
+            payment_hash = json_response_invoice['payment_hash']
+            invoice = json_response_invoice['payment_request']
+            expires_at = json_response_invoice['expires_at']
+
+            lightning_invoice = m.LightningInvoice(
+                order_id=order_id,
+                invoice=invoice,
+                payment_hash=payment_hash,
+                price=sats,
+                expires_at=expires_at         # "2023-08-03T08:52:07.473598061Z"
+            )
+            db.session.add(lightning_invoice)
+            db.session.commit()
+
+            return json_response_invoice
+
+        elif (response_invoice.status_code == 401):
+            app.logger.info(f"create_invoice - LndHub API error. Probably the token expired!")
+
+            self.auth_header = self.get_login_token()
+            return self.create_invoice(order_id, sats)
```

Comment:
Recursive call here looks dangerous. How if it never stops recursion? Can't we move this retry in the caller?

Reply:
I would really just have a method on LightningInvoice model that does the thing and maybe a retry... and not touch the model from here...

Reply:
You're right. It's ugly. I'll fix it.